### PR TITLE
[subset CFF] Fix de-subroutinizing bug when subroutines contain hints

### PR DESCRIPTION
**Pull request by readroberts**
Fixes issue 1493

The  code was skipping executing a subroutine if it had already been desubroutinized. The initial set of vstemhm and hstemhm operators and values may be in a subroutine. If a charstring is being executed which calls such subroutines, they still need to be executed in order to count the number of hint values seen, so that the byte length of the hintmask can be calculated.

I fixed this bug by executing subroutines even if they have already been desubroutinized, as long  as  (we don't know yet if we are doing hintmasks) or ( we do need a hintmask, but have not yet seen it).

### CODE REVIEW

File: Lib/fontTools/subset/cff.py [hunk 1]
```diff
@@ -352,28 +352,45 @@ def processSubr(self, index, subr):
...
 
 	def __init__(self, localSubrs, globalSubrs, private=None):
-		psCharStrings.SimpleT2Decompiler.__init__(self,
-							  localSubrs,
-							  globalSubrs, private)
+		psCharStrings.SimpleT2Decompiler.__init__(self, localSubrs, globalSubrs,
+												private)
 
 	def execute(self, charString):
+		self.need_hintcount = True  # until proven otherwise
```
Comment (cjchapman) [inline Lib/fontTools/subset/cff.py]: I suggest moving the initialization of `need_hintcount` into the `__init__()` method.

File: Lib/fontTools/subset/cff.py [hunk 2]
```diff
@@ -401,9 +418,24 @@ def op_callgsubr(self, index):
 		psCharStrings.SimpleT2Decompiler.op_callgsubr(self, index)
 		self.processSubr(index, subr)
 
+	def stop_hint_count(self, index=None):
```
Comment (cjchapman) [inline Lib/fontTools/subset/cff.py]: I think you can remove `index` from the parameter list as it's not used in this method, nor in the calls to this method.

File: Lib/fontTools/subset/cff.py (hunk 1 above)
Comment (readroberts) [reply Lib/fontTools/subset/cff.py]: The problem with that is that the class instance can be re-used for more than one charstring. Running it on one charstring will leave self.need_hintcount set to False, and it needs to be True at the start of execute(). I could set the value in init, and then extend the parent class's reset() method to set this to True. However, since this class is pretty specialized, I chose to set close to were it is used.

File: Lib/fontTools/subset/cff.py (hunk 2 above)
Comment (readroberts) [reply Lib/fontTools/subset/cff.py]: Actually, it is used. All of the functions in the stop_hintcount_ops list are called with two arguments: self , and index. stop_hint_count() needs accept the second argument, even though it isn't used in this functions. These functions are all called by  misc.psCharstrings.SimpleT2Decompiler.execute(), with the same handler call.

File: Lib/fontTools/subset/cff.py (hunk 1 above)
Comment (cjchapman) [reply Lib/fontTools/subset/cff.py]: OK.

File: Lib/fontTools/subset/cff.py (hunk 2 above)
Comment (cjchapman) [reply Lib/fontTools/subset/cff.py]: OK.

**Comment by anthrotype** [comment]
when calling the subsetter with `--desubroutinize` as well as with `--no-hinting`, we still get AttributeError for `GlobalSubrs` inside `remove_unused_subroutines` method, which is called at the end of `remove_hinting` method. This is because the `desubroutinize` method deletes the attribute on the FontDict.
This patch fixes this. Do you mind applying this and also add a little test case for using the two options in combination?

```diff
diff --git a/Lib/fontTools/subset/cff.py b/Lib/fontTools/subset/cff.py
index ee6048a0..9dc099b8 100644
--- a/Lib/fontTools/subset/cff.py
+++ b/Lib/fontTools/subset/cff.py
@@ -486,9 +486,7 @@ def desubroutinize(self):
                        decompiler.execute(c)
                        c.program = c._desubroutinized
                        del c._desubroutinized
-               # Delete All the Subrs!!!
-               if font.GlobalSubrs:
-                       del font.GlobalSubrs
+               # Delete all the local subrs
                if hasattr(font, 'FDArray'):
                        for fd in font.FDArray:
                                pd = fd.Private
@@ -502,6 +500,7 @@ def desubroutinize(self):
                                del pd.Subrs
                        if 'Subrs' in pd.rawDict:
                                del pd.rawDict['Subrs']
+       # as well as the global subrs
        cff.GlobalSubrs.clear()
```

Note that it is not necessary to call `del font.GlobalSubrs` to delete the global subroutines, since the attribute is shared by all the font dicts in the CFFFontSet and we call `clear()` at the end of `desubroutinize` to set it empty.

File: Lib/fontTools/subset/cff.py (hunk 2 above)
Comment (anthrotype) [reply Lib/fontTools/subset/cff.py]: if it's unused, it's clearer if you just use `*args, **kwargs` then

File: Lib/fontTools/subset/cff.py [hunk 3]
```diff
@@ -401,9 +418,24 @@ def op_callgsubr(self, index):
...
+	def op_hintmask(self, index):
+		psCharStrings.SimpleT2Decompiler.op_hintmask(self, index)
+		if self.need_hintcount:
+			self.stop_hint_count()
+
 	def processSubr(self, index, subr):
 		cs = self.callingStack[-1]
-		cs._patches.append((index, subr._desubroutinized))
+		if not hasattr(cs, '_desubroutinized'):
+			cs = self.callingStack[-1]
```
Comment (anthrotype) [inline Lib/fontTools/subset/cff.py]: this line 437 is repeated two lines above

File: Lib/fontTools/subset/cff.py [hunk 4]
```diff
@@ -352,28 +352,45 @@ def processSubr(self, index, subr):
...
+		for op_name in stop_hintcount_ops:
+			setattr(self, op_name, self.stop_hint_count)
+
 		if hasattr(charString, '_desubroutinized'):
+			if self.need_hintcount and self.callingStack:
+				try:
+					psCharStrings.SimpleT2Decompiler.execute(self, charString)
+				except StopHintCountEvent:
+					del self.callingStack[-1]
+				return
```
Comment (anthrotype) [inline Lib/fontTools/subset/cff.py]: this `return` statement is also unnecessary, the following line does the same

File: Lib/fontTools/subset/cff.py [hunk 5]
```diff
@@ -352,28 +352,45 @@ def processSubr(self, index, subr):
...
 		elif subr_hints.status == 0:
 			hints.deletions.append(index)
 
 		hints.status = max(hints.status, subr_hints.status)
 
+class StopHintCountEvent(Exception):
+	pass
+
+
+stop_hintcount_ops = ("op_hstem", "op_vstem", "op_rmoveto", "op_hmoveto",
```
Comment (anthrotype) [inline Lib/fontTools/subset/cff.py]: i'd make this `stop_hintcount_ops` a class variable, it's only used within `_DesubroutinizingT2Decompiler`

File: Lib/fontTools/subset/cff.py (hunk 2 above)
Comment (readroberts) [reply Lib/fontTools/subset/cff.py]: Agreed.

File: Lib/fontTools/subset/cff.py (hunk 3 above)
Comment (readroberts) [reply Lib/fontTools/subset/cff.py]: Right.

File: Lib/fontTools/subset/cff.py (hunk 4 above)
Comment (readroberts) [reply Lib/fontTools/subset/cff.py]: Right.

File: Lib/fontTools/subset/cff.py (hunk 5 above)
Comment (readroberts) [reply Lib/fontTools/subset/cff.py]: Right. Thanks for the review! Will fix all these, including the patch for the GlobalSubrs bug.

Comment (readroberts) [reply Lib/fontTools/subset/cff.py]: @anthrotype I see that there is already a test, subset_test.py::test_no_hinting_desubroutinize_CFF, which should encounter the AttributeError for `GlobalSubrs`. I can't reproduce this bug, either with this test, or using the command-line "pyftsubset Lobster.subset.otf  --gids=0-11 --desubroutinize --no-hinting". How did you reproduce this problem?

Comment (readroberts) [reply Lib/fontTools/subset/cff.py]: @anthrotype Never mind - I realize that the test needs a font with a non-empty GlobalSubrs. Have test case, will add.

File: Lib/fontTools/subset/cff.py [hunk 6]
```diff
@@ -470,6 +498,7 @@ def desubroutinize(self):
 				del pd.Subrs
 			if 'Subrs' in pd.rawDict:
 				del pd.rawDict['Subrs']
+	# as well as the global subrs
```
Comment (behdad) [inline Lib/fontTools/subset/cff.py]: How is this change relevant to the issue being fixed?

Comment (anthrotype) [reply Lib/fontTools/subset/cff.py]: I asked Read to apply that patch since he was working on related issues with the —desubroutinize option:
https://github.com/fonttools/fonttools/pull/1499#issuecomment-463309891
